File: Native/Attribute.cpp
#include "slang.h"
#include "slang-com-ptr.h"
#include "slang-com-helper.h"
#include "Attribute.h"


Native::Attribute::Attribute(void* native)
{
	m_native = (slang::Attribute*)native;
}

char const* Native::Attribute::getName()
{
	return m_native->getName();
}
uint32_t Native::Attribute::getArgumentCount()
{
	return m_native->getArgumentCount();
}
Native::TypeReflection* Native::Attribute::getArgumentType(uint32_t index)
{
	if (!m_argumentType)
		m_argumentType = new TypeReflection(m_native->getArgumentType(index));
	return m_argumentType;
}
SlangResult Native::Attribute::getArgumentValueInt(uint32_t index, int* value)
{
	return m_native->getArgumentValueInt(index, value);
}
SlangResult Native::Attribute::getArgumentValueFloat(uint32_t index, float* value)
{
	return m_native->getArgumentValueFloat(index, value);
}
const char* Native::Attribute::getArgumentValueString(uint32_t index, size_t* outSize)
{
	return m_native->getArgumentValueString(index, outSize);
}

File: src/Slang.Sdk/Layer3-Pretty/Reflection-API/Attribute.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Slang.Sdk.Collections;

namespace Slang.Sdk
{
    public class Attribute : Reflection, IEquatable<Attribute>,
        IComposition<Type>
    {
        #region Definition
        public override Reflection? Parent { get; }
        internal override Binding.AttributeReflection Binding { get; }

        internal Attribute(Reflection parent, Binding.AttributeReflection binding)
        {
            Parent = parent;
            Binding = binding;
        }
        #endregion

        #region IComposition<Type> (Argument Types)
        uint IComposition<Type>.Count => Binding.GetArgumentCount();
        Type IComposition<Type>.GetByIndex(uint index) =>
            new Type(this, Binding.GetArgumentType(index));

        #endregion

        #region Collections

        SlangCollection<Type>? _ArgumentTypes;
        public SlangCollection<Type> ArgumentTypes => _ArgumentTypes ??= new SlangCollection<Type>(this);

        #endregion

        #region Pretty
        public string? Name => Binding.GetName();

        public int GetArgumentValueInt(uint index) => Binding.GetArgumentValueInt(index);

        public float GetArgumentValueFloat(uint index) => Binding.GetArgumentValueFloat(index);

        public string? GetArgumentValueString(uint index) => Binding.GetArgumentValueString(index);

        #endregion

        #region Equality
        public bool Equals(Attribute? other)
        {
            return this == other;
        } 
        #endregion
    }
}
